designs/odcr.md

```diff
@@ -0,0 +1,204 @@
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Proposed Solution](#proposed-solution)
+    - [Supporting associating Capacity Reservations to EC2NodeClass](#supporting-associating-capacity-reservations-to-ec2nodeclass)
+    - [Supporting new capacity-type "capacity-reservation" for NodePool](#supporting-new-capacity-type-capacity-reservation-for-nodepool)
+      - [only allow capacity-reservations](#only-allow-capacity-reservations)
+      - [capacity-reservation-fleet (falling back to on-demand if capacity-reservation not available)](#capacity-reservation-fleet-falling-back-to-on-demand-if-capacity-reservation-not-available)
+    - [Launching Nodes into Capacity Reservation](#launching-nodes-into-capacity-reservation)
+      - [All Launch Templates are associated with the specified Capacity Reservation](#all-launch-templates-are-associated-with-the-specified-capacity-reservation)
+      - [Pricing and consolidation](#pricing-and-consolidation)
+        - [Provisioning](#provisioning)
+        - [Consolidating Capacity Reserved Instances](#consolidating-capacity-reserved-instances)
+        - [Consolidating into Capacity Reserved Instances](#consolidating-into-capacity-reserved-instances)
+      - [Labels](#labels)
+    - [Failed to launch Nodes into Capacity Reservation](#failed-to-launch-nodes-into-capacity-reservation)
+
+## Background
+
+In AWS [ODCR](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-capacity-reservations.html) allows users to reserve compute capacity to mitigate the risk of 
+getting on-demand capacity. This is very helpful during seasonal holidays where higher traffic is expected or for reserving highly-desired instance types, like the
+`p5.48xlarge` or other large GPU instance types.
+
+### Capacity Reservations
+
+Each Capacity Reservation is defined with:
+
+- The Availability Zone in which to reserve the capacity
+- The number of instances for which to reserve capacity
+- The instance attributes, including the instance type, tenancy, and platform/OS
+- Instance match criteria
+  - Targeted -- only accept instances that matches all attributes + explicitly targeted the capacity reservation
+  - Open -- if capacity reservation accepts all instances that matches all attributes
+
+AWS also supports grouping Capacity Reservation into Capacity Reservation groups. 
+Both these entities are supported in Launch Template's CapacityReservationTarget [definitions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationtarget.html).
+
+## Goals
+
+- Support associating targeted ODCRs to EC2NodeClass
+- Define Karpenter's behavior when using new Capacity Type "capacity-reservation" in NodePool
+- Define Karpenter's behavior when encountering errors when attempting to launch nodes into Capacity Reservation
+- Define Karpenter's behavior when consolidating nodes to Capacity Reservation when available
+
+## Non-Goals
```

Comment:
May want to add Capacity Blocks to this list. I think Karpenter should support Capacity Blocks, but you may want to separate that out (or else explicitly call it out in the goals).

Reply:
Is CBR part of ODCR? The doc's title is specific to ODCR so I didn't think we would include any info around CBR

Reply:
CBR is a separate reservation process from ODCR. I think the main point is just being explicitly clear about what this RFC is addressing, and also clear in the terminology used throughout to avoid any confusion between the two forms. For example `Capacity Reservation` is generically used throughout - its not specific enough to disambiguate which form of capacity reservation

Reply:
Capacity Blocks are a type of Capacity Reservation. It makes more sense for the API to be around "capacity reservations" (which is what the EC2 APIs are around) versus "ODCRs", which is a short hand for a capacity reservation that is typed to on-demand. I see that your RFC calls out "capacityReservationSelectorTerms", which does not sound ODCR specific to me. As a result, does not seem to me to be obvious that Capacity Blocks are out of scope, hence my comment that it should probably be in your non-goal list and is probably something to think about in the design (even if you don't implement it now) because I'm guessing some folks will be interested in Capacity Blocks support.

Reply:
I suspect that we will make capacity blocks in scope of the same capacityReservationSelectorTerms when we add support for them.